File: src/connectors/claude-ai.ts
import fs from "fs/promises";
import path from "path";
import fg from "fast-glob";
import type { RawConversation, ConnectorConfig } from "../kb/types.js";
import { BaseConnector } from "./base.js";

export class ClaudeAIConnector extends BaseConnector {
  name = "claude_ai";
  private exportPath: string;

  constructor(config: ConnectorConfig) {
    super(config);
    this.exportPath = config.export_path as string || "";
  }

  async validate(): Promise<boolean> {
    if (!this.exportPath) return false;
    try {
      await fs.access(this.exportPath);
      return true;
    } catch {
      return false;
    }
  }

  async fetch(since?: Date): Promise<RawConversation[]> {
    if (!this.exportPath) return [];

    // Claude.ai export is a directory of JSON files
    const files = await fg("*.json", {
      cwd: this.exportPath,
      absolute: true,
    });

    const results: RawConversation[] = [];

    for (const file of files) {
      try {
        const stat = await fs.stat(file);
        if (since && stat.mtime < since) continue;

        const raw = await fs.readFile(file, "utf-8");
        const conv = JSON.parse(raw);

        const messages = (conv.chat_messages || [])
          .map((m: { sender: string; text: string; created_at?: string }) => ({
            role: m.sender === "human" ? "user" as const : "assistant" as const,
            content: m.text || "",
            timestamp: m.created_at,
          }))
          .filter((m: { content: string }) => m.content.trim());

        if (messages.length > 0) {
          results.push({
            source: "claude_ai",
            id: path.basename(file, ".json"),
            messages,
            timestamp: stat.mtime.toISOString(),
            metadata: { file_path: file, name: conv.name },
          });
        }
      } catch {
        // skip
      }
    }

    return results;
  }
}

File: synesis/connectors/claude_ai.py
from __future__ import annotations

import glob
import json
from datetime import datetime
from pathlib import Path

from synesis.connectors.base import BaseConnector
from synesis.kb.types import ConversationMessage, RawConversation


class ClaudeAIConnector(BaseConnector):
    name = "claude_ai"

    def __init__(self, config: dict):
        super().__init__(config)
        self.export_path = config.get("export_path") or ""

    def validate(self) -> bool:
        if not self.export_path:
            return False
        return Path(self.export_path).exists()

    def fetch(self, since: str | None = None) -> list[RawConversation]:
        if not self.export_path:
            return []

        since_dt = datetime.fromisoformat(since) if since else None
        results: list[RawConversation] = []

        for json_file in glob.glob(str(Path(self.export_path) / "*.json")):
            path = Path(json_file)
            mtime = datetime.fromtimestamp(path.stat().st_mtime)
            if since_dt and mtime < since_dt:
                continue

            try:
                conv = json.loads(path.read_text(encoding="utf-8"))
                messages = [
                    ConversationMessage(
                        role="user" if m.get("sender") == "human" else "assistant",
                        content=m.get("text", ""),
                        timestamp=m.get("created_at"),
                    )
                    for m in conv.get("chat_messages", [])
                    if m.get("text", "").strip()
                ]

                if messages:
                    results.append(
                        RawConversation(
                            source="claude_ai",
                            id=path.stem,
                            messages=messages,
                            timestamp=mtime.isoformat(),
                            metadata={"file_path": str(path), "name": conv.get("name")},
                        )
                    )
            except (json.JSONDecodeError, KeyError):
                continue

        return results
